### ISSUE TITLE
filter_by() doesn't derive the entity from func calls, maybe other expressions

### ISSUE THREAD

**Issue by ThiefMaster** [sqlalchemy/sqlalchemy#6234]
**Describe the bug**
Using a callable with context as a default value breaks badly when triggering a query within it.

**Expected behavior**
No error, and correctly querying the default value, like it was the case before 1.4

**To Reproduce**

```py
from sqlalchemy import *
from sqlalchemy.orm import *


Base = declarative_base()



def _get_next_position(context):
    parent_id = context.current_parameters['parent_id']
    res = s.query(func.max(Category.position)).filter_by(parent_id=parent_id).one()
    return (res[0] or 0) + 1


class Category(Base):
    __tablename__ = 'categories'
    id = Column(Integer, primary_key=True)
    parent_id = Column(Integer, ForeignKey('categories.id'), index=True, nullable=True)
    position = Column(SmallInteger, nullable=False, default=_get_next_position)
    children = relationship('Category', primaryjoin=(id == remote(parent_id)),
                            backref=backref('parent', primaryjoin=(remote(id) == parent_id)))

    def __repr__(self):
        return f'<Category {self.id} @{self.position}>'


e = create_engine('postgresql:///test', echo=True)
Base.metadata.create_all(e)
s = Session(e)


root = Category()
Category(parent=root)
Category(parent=root)
s.add(root)
s.commit()

print(s.query(Category).all())

```

**Error**

```
2021-04-09 18:26:14,908 INFO sqlalchemy.engine.Engine select version()
2021-04-09 18:26:14,909 INFO sqlalchemy.engine.Engine [raw sql] {}
2021-04-09 18:26:14,909 INFO sqlalchemy.engine.Engine select current_schema()
2021-04-09 18:26:14,909 INFO sqlalchemy.engine.Engine [raw sql] {}
2021-04-09 18:26:14,910 INFO sqlalchemy.engine.Engine show standard_conforming_strings
2021-04-09 18:26:14,910 INFO sqlalchemy.engine.Engine [raw sql] {}
2021-04-09 18:26:14,910 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2021-04-09 18:26:14,910 INFO sqlalchemy.engine.Engine select relname from pg_class c join pg_namespace n on n.oid=c.relnamespace where pg_catalog.pg_table_is_visible(c.oid) and relname=%(name)s
2021-04-09 18:26:14,910 INFO sqlalchemy.engine.Engine [generated in 0.00008s] {'name': 'categories'}
2021-04-09 18:26:14,911 INFO sqlalchemy.engine.Engine COMMIT
2021-04-09 18:26:14,914 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2021-04-09 18:26:14,915 INFO sqlalchemy.engine.Engine ROLLBACK
Traceback (most recent call last):
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/sql/base.py", line 1156, in __getattr__
    return self._index[key]
KeyError: 'parent_id'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/sql/base.py", line 1580, in _entity_namespace_key
    return getattr(ns, key)
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/sql/base.py", line 1158, in __getattr__
    util.raise_(AttributeError(key), replace_context=err)
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 211, in raise_
    raise exception
AttributeError: parent_id

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1608, in _execute_context
    context = constructor(
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 972, in _init_compiled
    self._process_executesingle_defaults()
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 1840, in _process_executesingle_defaults
    val = self.get_insert_default(c)
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/dialects/postgresql/base.py", line 3028, in get_insert_default
    return super(PGExecutionContext, self).get_insert_default(column)
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 1785, in get_insert_default
    return self._exec_default(column, column.default, column.type)
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 1649, in _exec_default
    return default.arg(self)
  File "/home/adrian/dev/indico/py3/src/satest.py", line 11, in _get_next_position
    res = s.query(func.max(Category.position)).filter_by(parent_id=parent_id).one()
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/orm/query.py", line 1743, in filter_by
    clauses = [
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/orm/query.py", line 1744, in <listcomp>
    _entity_namespace_key(from_entity, key) == value
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/sql/base.py", line 1582, in _entity_namespace_key
    util.raise_(
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 211, in raise_
    raise exception
sqlalchemy.exc.InvalidRequestError: Entity namespace for "max(categories.position)" has no property "parent_id"

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/adrian/dev/indico/py3/src/satest.py", line 36, in <module>
    s.commit()
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 1415, in commit
    self._transaction.commit(_to_root=self.future)
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 829, in commit
    self._prepare_impl()
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 808, in _prepare_impl
    self.session.flush()
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 3233, in flush
    self._flush(objects)
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 3373, in _flush
    transaction.rollback(_capture_exception=True)
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/util/langhelpers.py", line 70, in __exit__
    compat.raise_(
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 211, in raise_
    raise exception
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 3333, in _flush
    flush_context.execute()
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/orm/unitofwork.py", line 450, in execute
    n.execute_aggregate(self, set_)
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/orm/unitofwork.py", line 744, in execute_aggregate
    persistence.save_obj(
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/orm/persistence.py", line 242, in save_obj
    _emit_insert_statements(
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/orm/persistence.py", line 1219, in _emit_insert_statements
    result = connection._execute_20(
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1520, in _execute_20
    return meth(self, args_10style, kwargs_10style, execution_options)
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/sql/elements.py", line 313, in _execute_on_connection
    return connection._execute_clauseelement(
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1389, in _execute_clauseelement
    ret = self._execute_context(
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1614, in _execute_context
    self._handle_dbapi_exception(
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1929, in _handle_dbapi_exception
    util.raise_(
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 211, in raise_
    raise exception
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1608, in _execute_context
    context = constructor(
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 972, in _init_compiled
    self._process_executesingle_defaults()
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 1840, in _process_executesingle_defaults
    val = self.get_insert_default(c)
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/dialects/postgresql/base.py", line 3028, in get_insert_default
    return super(PGExecutionContext, self).get_insert_default(column)
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 1785, in get_insert_default
    return self._exec_default(column, column.default, column.type)
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 1649, in _exec_default
    return default.arg(self)
  File "/home/adrian/dev/indico/py3/src/satest.py", line 11, in _get_next_position
    res = s.query(func.max(Category.position)).filter_by(parent_id=parent_id).one()
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/orm/query.py", line 1743, in filter_by
    clauses = [
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/orm/query.py", line 1744, in <listcomp>
    _entity_namespace_key(from_entity, key) == value
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/sql/base.py", line 1582, in _entity_namespace_key
    util.raise_(
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 211, in raise_
    raise exception
sqlalchemy.exc.StatementError: (sqlalchemy.exc.InvalidRequestError) Entity namespace for "max(categories.position)" has no property "parent_id"
[SQL: INSERT INTO categories (parent_id, position) VALUES (%(parent_id)s, %(position)s) RETURNING categories.id]
[parameters: [{'parent_id': None}]]

```

**Versions.**
 - OS: Linux
 - Python: 3.9.1
 - SQLAlchemy: 1.4.6
 - Database: postgres
 - DBAPI: psycopg2

**Comment by zzzeek** [comment]
heya.  

This is not related to the default or anything like that.   filter_by() doesn't know how to get at "parent_id" from the func.max() function.     This is sort of bug-ish and I can try fixing it, but easy enough just skip the filter_by:
```

print(
    s.query(func.max(Category.position)).where(Category.parent_id==1)
)

```

here's the full MCVE with the possibly buglike behavior:

```python
from sqlalchemy import Column
from sqlalchemy import func
from sqlalchemy import Integer
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import Session


Base = declarative_base()



class Category(Base):
    __tablename__ = 'categories'
    id = Column(Integer, primary_key=True)
    parent_id = Column(Integer)
    position = Column(Integer)

s = Session()

print(
    s.query(func.max(Category.position)).where(Category.parent_id==1)
)

print(
    s.query(func.max(Category.position)).filter_by(parent_id=1)
)


```

**Comment by zzzeek** [comment]
it's a bug.  but im out of time for today and i have to put 1.4.7 out, will try to see if i can hit it over the weekend.

**Comment by ThiefMaster** [comment]
Ah that workaround is fine. No need to `filter_by` here.

**Comment by zzzeek** [comment]
i was trying to find this issue in May, as it was reported again, couldn't find it.  this was fixed in 1.4.12 dupe of #6414.
